src/lib/core/datetime.c

```diff
@@ -0,0 +1,78 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2021, Tarantool AUTHORS, please see AUTHORS file.
+ */
+
+#include <assert.h>
+#include <limits.h>
+#include <string.h>
+#include <time.h>
+
+#include "c-dt/dt.h"
+#include "trivia/util.h"
+#include "datetime.h"
+
+void
+datetime_now(struct datetime *now)
+{
+	struct timeval tv;
+	gettimeofday(&tv, NULL);
+	now->epoch = tv.tv_sec;
+	now->nsec = tv.tv_usec * 1000;
+
+	struct tm tm;
+	localtime_r(&tv.tv_sec, &tm);
+	now->tzoffset = tm.tm_gmtoff / 60;
+}
+
+/**
+ * NB! buf may be NULL, and we should handle it gracefully, returning
+ * calculated length of output string
+ */
+int
+datetime_to_string(char *buf, int len, const struct datetime *date)
```

Comment:
1. struct datetime should be the first argument (OOP)?
2. why don't you use `strftime` here?

Reply:
1. There are number of `*_snprint*` functions which are being used with `SNPRINT()` macro for stringization of various Tarantool types (e.g. `tuple_snprint`, `key_def_snprint_parts`, `histogram_snprint`, etc.) they use this unusual order;
2. Because of 2 reasons:
  * It's 10x times faster when there is no need to parse format;
  * we need full control of a resultant value, regardless of a strftime formats supported. We need full control even for non-C contexts (i.e. for yaml or json).

Reply:
I've reordered arguments, also reusing `size_t`/`ssize_t` types appropriately (wherever I need unsigned or signed integer for size values):

```diff
diff --git a/src/lib/core/datetime.c b/src/lib/core/datetime.c
index f9c0b209e..e3d9e013a 100644
--- a/src/lib/core/datetime.c
+++ b/src/lib/core/datetime.c
@@ -70,8 +70,8 @@ datetime_strftime(const struct datetime *date, char *buf, size_t len,
 /* NB! buf may be NULL, and we should handle it gracefully, returning
  * calculated length of output string
  */
-int
-datetime_to_string(char *buf, int len, const struct datetime *date)
+size_t
+datetime_to_string(const struct datetime *date, char *buf, ssize_t len)
 {
 	int offset = date->tzoffset;
 	int64_t rd_seconds = (int64_t)date->epoch + offset * 60 +
@@ -89,7 +89,7 @@ datetime_to_string(char *buf, int len, const struct datetime *date)
 	second = rd_seconds % 60;
 	nanosec = date->nsec;
 
-	int sz = 0;
+	size_t sz = 0;
 	SNPRINT(sz, snprintf, buf, len, "%04d-%02d-%02dT%02d:%02d:%02d",
 		year, month, day, hour, minute, second);
 	if (nanosec != 0) {
diff --git a/src/lib/core/datetime.h b/src/lib/core/datetime.h
index b1e9d8f9e..9bfd2f533 100644
--- a/src/lib/core/datetime.h
+++ b/src/lib/core/datetime.h
@@ -55,13 +55,13 @@ struct datetime {
 
 /**
  * Convert datetime to string using default format
+ * @param date source datetime value
  * @param buf output character buffer
  * @param len size of output buffer
- * @param date sourceadatetime value
  * @retval length of a resultant text
  */
-int
-datetime_to_string(char *buf, int len, const struct datetime *date);
+size_t
+datetime_to_string(const struct datetime *date, char *buf, ssize_t len);
 
 /**
  * Convert datetime to string using default format provided
diff --git a/src/lua/datetime.lua b/src/lua/datetime.lua
index b4d64617c..c0e1c3051 100644
--- a/src/lua/datetime.lua
+++ b/src/lua/datetime.lua
@@ -56,7 +56,7 @@ int     tnt_dt_dom          (dt_t dt);
 size_t tnt_dt_parse_iso_zone_lenient(const char *str, size_t len, int *offset);
 
 /* Tarantool functions - datetime.c */
-int    datetime_to_string(char *buf, int len, const struct datetime * date);
+size_t datetime_to_string(const struct datetime * date, char *buf, ssize_t len);
 size_t datetime_strftime(const struct datetime *date, char *buf, uint32_t len,
                          const char *fmt);
 void   datetime_now(struct datetime *now);
@@ -352,7 +352,7 @@ local function datetime_tostring(self)
     check_date(self, 'datetime.tostring()')
 
     local buff = ffi.new('char[?]', TOSTRING_BUFSIZE)
-    local len = builtin.datetime_to_string(buff, TOSTRING_BUFSIZE, self)
+    local len = builtin.datetime_to_string(self, buff, TOSTRING_BUFSIZE)
     assert(len < TOSTRING_BUFSIZE)
     return ffi.string(buff)
 end
```